fix: throw error on sub-day generate_series increments

tshauck opened the pull request:
## Which issue does this PR close?

closes #11823

## Rationale for this change

Currently, when trying to generate a series of dates with increments less than a day, the query will hang due to loss of precision (the step size becomes 0 relative to the Date32).

## What changes are included in this PR?

* Raise an error if attempting increment by less than a day
* Refactor to use a `ListBuilder`
* A few minor quick-fixes

## Are these changes tested?

Yes, added a test.

## Are there any user-facing changes?

Minor in that queries will now error when they would've hung forever.

alamb reviewed (commented):
Thank you very much @tshauck  -- I think this code looks much nicer 👌  -- thank you

While reviewing this PR I had some more ideas for tests, so I took the liberty of writing them and pushing a few to your branch.

During this process, I actually think I found a regression introduced by this PR. Specifically, when generating intervals from arrays. If you try and run the tests in https://github.com/apache/datafusion/pull/11921 on this PR they all generate internal errors.

tshauck commented:
@alamb I rebased and updated the code to avoid those internal errors. I think root cause was me using `.append_value` w/ `.append` leading to more rows being returned than expected.

alamb reviewed (approved):
Thanks @tshauck  -- this looks great to me
